### PR TITLE
WIP: Remove invalid merge_repeated option from CTC beam decoder

### LINKED ISSUE #9550: ctc_greedy_decoder inconsistent with ctc_beam_search_decoder
The following extract from the the `ctc_beam_search_decoder` documentation seems to be misleading:
"The `ctc_greedy_decoder` is a special case of the `ctc_beam_search_decoder` with `top_paths=1` and `beam_width=1` (but that decoder is faster for this special case)."

Instead, the following results can be observed: 

| Decoding "AA<ctc_blank>AA" using | `merge_repeated=True` | `merge_repeated=False` |
| --- | --- | --- |
| `tf.nn.ctc_beam_search_decoder(top_paths=1, beam_width=1)` | "A"   | "AA" |
| `tf.nn.ctc_greedy_decoder()` | "AA" | "AAAA" |

To reproduce:

```python
import numpy as np
import tensorflow as tf
from unittest import TestCase


class CtcDecodersTest(TestCase):
    def test(self):
        def decode_greedily(beam_search: bool, merge_repeated: bool):
            aa_ctc_blank_aa_logits = tf.constant(np.array([[[1.0, 0.0]], [[1.0, 0.0]], [[0.0, 1.0]],
                                                    [[1.0, 0.0]], [[1.0, 0.0]]], dtype=np.float32))
            sequence_length = tf.constant(np.array([5], dtype=np.int32))

            (decoded_list,), log_probabilities = \
                tf.nn.ctc_beam_search_decoder(inputs=aa_ctc_blank_aa_logits,
                                              sequence_length=sequence_length,
                                              merge_repeated=merge_repeated,
                                              beam_width=1) \
                    if beam_search else \
                    tf.nn.ctc_greedy_decoder(inputs=aa_ctc_blank_aa_logits,
                                             sequence_length=sequence_length,
                                             merge_repeated=merge_repeated)

            return list(tf.Session().run(tf.sparse_tensor_to_dense(decoded_list)[0]))

        self.assertEqual([0], decode_greedily(beam_search=True, merge_repeated=True))
        self.assertEqual([0, 0], decode_greedily(beam_search=True, merge_repeated=False))
        self.assertEqual([0, 0], decode_greedily(beam_search=False, merge_repeated=True))
        self.assertEqual([0, 0, 0, 0], decode_greedily(beam_search=False, merge_repeated=False))
```

This is confusing and probably not intended.

How to solve this:

- Adapt the documentation or
- Adapt the `ctc_beam_search_decoder` implementation to that of `ctc_greedy_decoder` or vice versa. Both directions would cover my use case ("AA"), this decision would depend on which of the other behaviors ("A" or "AAAA") is needed and which of them could be dropped.

### System information
OSX 12.4, TensorFlow 1.1.0 CPU from binary

### PR DESCRIPTION
**Pull request by ryanleary**
The CTC beam decoding implicitly collapses repeated characters as part of calculating the optimal path (i.e. 'AAA' will contribute probability mass through the path 'A').

So the correct CTC decoding behavior occurs when merge_repeated=False. In this case, it DOES merge repeated characters. The merge_repeated flag, when true, will merge repeated characters after characters have already been merged/blank symbols removed. As it stands now, the behavior is extremely misleading.

This PR removes the misleading parameter from the ctc beam search op.

Closes #9550.

### CODE REVIEW

**Comment by ryanleary** [comment]
I signed it!

**Comment by ryanleary** [comment]
Bueller?

**Comment by ebrevdo** [comment]
This breaks backwards compatibility (and all users who currently have an old graph built using the ctc loss, regardless of the argument's value), so we definitely cannot accept this PR as is!

Can you instead update the documentation, making it very clear what the arguments mean - and adding a warning about use cases?  Julian's table looks like a very good item to add to the docs.

**Comment by ryanleary** [comment]
Can you enumerate a use case where `merge_repeated=True` is actually valid? I recognize it's a breaking change, but the current behavior, IMO, is a bug.

**Comment by ryanleary** [comment]
Happy to update docs in the interim.

**Comment by ebrevdo** [comment]
Regardless of the current behavior; this PR, as is, introduces a worse bug:
namely that existing graphs will crash in the new version of the C++
runtime.  That's against TF's backwards compatibility guarantees.

You are welcome to add a deprecation warning if you strongly believe this
argument to be a bug -- we will review it.  There's even a decorator that
allows deprecating function arguments in TF.

**Comment by ryanleary** [comment]
Got it. Definitely agree that this PR is a WIP, but wanted to get some feedback. I do believe that the argument shouldn't exist (and would like to encourage feedback/discussion on that contention).

If there's a more appropriate way to deprecate the kwarg (and apparently there is), I'll look into doing using that mechanism for achieving the same end goal.

**Comment by rmlarsen** [comment]
@ryanleary do you intend to continue working on this PR or can we close it?

**Comment by standy66** [comment]
@ryanleary any update on this PR?

**Comment by ryanleary** [comment]
@rmlarsen Could you send me a pointer to an appropriate way to deprecate a kwarg? I think the argument really should be removed in future versions.

**Comment by protoget** [comment]
Re @ryanleary  Check this:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/rnn/python/ops/gru_ops.py#L135

**Comment by ebrevdo** [comment]
yes; feel free to add the deprecation decorator!

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by tensorflowbutler** [comment]
It has been 58 days that this pull-request has stalled. Please create a new pull-request with the requested changes.

**Comment by ebrevdo** [comment]
I'm about to push a v2 with this argument removed.
